feat: introduce from_str_with_options() to support custom stylesheets

bolinfest opened the pull request.

bolinfest committed:
feat: introduce from_str_with_options() to support custom stylesheets

joshka commented on tui-markdown/src/options.rs:
```diff
@@ -0,0 +1,40 @@
+#[derive(Clone)]
+pub struct Options<S: StyleSheet = BuiltinStyleSheet> {
+    /// The [`StyleSheet`] implementation that will be consulted every time the
+    /// renderer needs a colour choice.
```
nit: Here and elsewhere, use the US spelling of color. I'm Australian (currently in the US), but I generally use the US spelling regardless on programming terms like this because it's easier to be consistent across docs and code.

```suggestion
    /// renderer needs a color choice.
```

joshka commented on tui-markdown/src/options.rs:
```diff
@@ -0,0 +1,40 @@
+    /// renderer needs a colour choice.
+    pub styles: S,
+
+    /// Wrap text at this many columns.  `None` means *no wrapping* (current
+    /// behaviour).
```
Current behavior refers to before the Option type was introduced, but won't have any relevance to someone who's looking at the docs for the first time, so it doesn't need to be stated.

```suggestion
    /// Wrap text at this many columns.
    ///
    /// The default value is `None` which indicates that there is no wrapping.
```

joshka commented on tui-markdown/src/options.rs:
```diff
@@ -0,0 +1,40 @@
+    /// behaviour).
+    pub wrap_width: Option<u16>,
+
+    /// Whether to show line numbers inside fenced code blocks.  *Not yet
+    /// implemented* – but the flag is here so that the field layout stays
+    /// stable once the feature lands.
+    pub show_line_numbers: bool,
```
This is a field where I'd be inclined to go with the non_exhaustive attribute rather than adding it here. WDYT?

joshka replied:
I removed these for now

joshka commented on tui-markdown/src/lib.rs:
```diff
@@ -36,16 +36,24 @@
 
+// Public-facing theming and configuration APIs --------------------------------
+mod style_sheet;
+pub use style_sheet::{BuiltinStyleSheet, DefaultStyleSheet, StyleSheet};
```
The Builtin vs Default distinction is a little confusing. Can you expand on why this is necessary / what it solves? Without understanding this, I'd intuitively suggest perhaps keep only one of these.

joshka commented on tui-markdown/src/lib.rs:
```diff
@@ -224,14 +253,15 @@ where
-            HeadingLevel::H4 => styles::H4,
-            HeadingLevel::H5 => styles::H5,
-            HeadingLevel::H6 => styles::H6,
+        let numeric_level: u8 = match level {
+            HeadingLevel::H1 => 1,
+            HeadingLevel::H2 => 2,
+            HeadingLevel::H3 => 3,
+            HeadingLevel::H4 => 4,
+            HeadingLevel::H5 => 5,
+            HeadingLevel::H6 => 6,
         };
+        let style = self.styles.heading(numeric_level);
```
The enum's variants have numeric discriminants that are fine to use (https://docs.rs/pulldown-cmark/latest/pulldown_cmark/enum.HeadingLevel.html)
```suggestion
        let style = self.styles.heading(level as u8);
```

joshka commented on tui-markdown/src/style_sheet.rs:
```diff
@@ -0,0 +1,89 @@
+/// choice we add a new getter here.  The default implementation maintains full
+/// backward-compatibility with the styles that lived in the old `mod styles`.
+pub trait StyleSheet: Clone + Send + Sync + 'static {
+    /// Style for a Markdown heading.  `level` is one-based (`1` for `# H1`, …).
```
```suggestion
    /// Style for a Markdown heading.
    ///
    /// `level` is one-based (`1` for `# H1`, …).
```

joshka commented on tui-markdown/src/style_sheet.rs:
```diff
@@ -0,0 +1,89 @@
+    fn blockquote(&self) -> Style;
+}
+
+/* ------------------------------------------------------------------------- */
+/*  Built-in style sheet – identical to the previous hard-coded colours       */
+/* ------------------------------------------------------------------------- */
```
nit: I prefer to avoid borders in code like this. Fold it into the doc comments.

joshka commented on tui-markdown/src/style_sheet.rs:
```diff
@@ -0,0 +1,89 @@
+impl StyleSheet for DefaultStyleSheet {
+    fn heading(&self, level: u8) -> Style {
+        match level {
+            1 => Style::new()
+                .bg(Color::Cyan)
+                .add_modifier(Modifier::BOLD)
+                .add_modifier(Modifier::UNDERLINED),
+            2 => Style::new()
+                .fg(Color::Cyan)
+                .add_modifier(Modifier::BOLD),
+            3 => Style::new()
+                .fg(Color::Cyan)
+                .add_modifier(Modifier::BOLD)
+                .add_modifier(Modifier::ITALIC),
+            4 => Style::new()
+                .fg(Color::LightCyan)
+                .add_modifier(Modifier::ITALIC),
+            5 => Style::new()
+                .fg(Color::LightCyan)
+                .add_modifier(Modifier::ITALIC),
+            _ => Style::new()
+                .fg(Color::LightCyan)
+                .add_modifier(Modifier::ITALIC),
```
use [Stylize](https://docs.rs/ratatui/latest/ratatui/style/trait.Stylize.html) here to make this concise. E.g.:
```rust
1 => Style::new().cyan().bold().underline(),
```

joshka reviewed (commented):
Lots of nits and a couple of main comments:

First up, thanks for making this configurable. It's something that was on my todo list for sometime when I had some time and energy to do it, so I'm always happy to see a contribution that pushes things forward.

In my initial thinking about this, I'd anticipated that Styling like this would have generally just been a struct with values rather than a trait. This obviously means that there would be more fields for headings to deal with. I'm curious whether you'd thought about any downsides with that approach that using a trait instead solves neatly. The only thing I can really think of is that there's some sort of dynamic values being returned, but I'm not sure that I can think of a good use case for that. This is not a hard blocker, but it's an interesting question, and we face a similar question with Ratatui core when thinking about how to implement theming, so it's worth getting some ideas / understanding here.

The main other questions I have:
- non_exhaustive
- Default vs BuiltIn

Overall I'm mostly fine with the approach you've taken - I mostly want to understand a few things to work out whether it's the best we can do, and would anticipate pushing forward with this once those are resolved.

joshka commented:
@bolinfest are you keen to get this over the line?

joshka committed:
fix: lib.rs review comments

joshka committed:
fix: other review comments / nits

joshka committed:
fix: doc updates, remove Builtin (just use DefaultStylesheet)

joshka committed:
fix: Add Options constructor and docs. Remove wrap / line number options for now

joshka commented:
For now, I changed this so that there's a constructor that accepts a stylesheet, I'd anticipate that adding more options would add builder style methods similar to the way that we do these in Ratatui (Matklad describes this as the builder-lite pattern in https://matklad.github.io/2022/05/29/builder-lite.html).

I've removed the wrapping and line number options, and will add them when that feature is added.

joshka committed:
fix(renderer): restore heading levels

Restore 1-based heading levels so stylesheet themes map to H1–H6 as before. Also remove
unused legacy styles and align style docs/imports with ratatui_core to keep clippy clean.
